### PR TITLE
Add Exception.blame/3 and Exception.blame_mfa/3

### LINKED ISSUE #4171: Suggesting the correct patterns to match
Hello,
This is something that would have saved me personally a lot of time.
For function headers, wouldn't it be extremely useful if the compiler/parser suggested the intended pattern that the code was trying to match when raising a `FunctionClauseError`?

For example, I wrote the following function:
`def render("template.json", %{group: group})`

I believed (wrongly) that the second argument was going to be %{group: some_group_variable}.
Especially that I did have other related functions that worked fine with this pattern.
I kept getting a FunctionClauseError and I didn't know why (my fault eventually, of course).

Second scenario, I misspelled an argument in a pattern. The function was (phoenix):
`def video_path(conn, :update, ...)`

And I called:
`video_path(conn, :udpate, ...)`

I went crazy here. I saw the typo in its correct form until I caught it after a few hours after I had some "away" time.

I think it would be very valuable to do something like:

```
** (FunctionClauseError) no function clause matching in Rumbl.Router.Helpers.video_path/4
** Provided pattern: video_path(conn, :udpate)
** Valid patterns:
        video_path(&1, :index)
        video_path(&1, :update)
        video_path(&1, :delete)
```

Or something similar to this.

### LINKED ISSUE #5737: Improve no function clause error messages
Today we say:

```iex
iex(1)> Regex.split "dsadsa", 321
** (FunctionClauseError) no function clause matching in Regex.split/3
    (elixir) lib/regex.ex:397: Regex.split("dsadsa", 321, [])
```

I believe we could improve this message. One proposal:

```iex
iex(1)> Regex.split "dsadsa", 321
** (FunctionClauseError) there is no definition of Regex.split/3 that expects the arguments at positions:

  1) "dsadsa"
  2) 321
  3) []

    (elixir) lib/regex.ex:397: Regex.split/3
```

The idea is to move the arguments themselves up to the error message and remove them from the stacktrace. I am not sure if the positions syntax above is good enough™ but it should be a good starting point for the discussion.

We could use asterisks instead of a numbered list as well:

```iex
iex(1)> Regex.split "dsadsa", 321
** (FunctionClauseError) there is no definition of Regex.split/3 that expects the following arguments:

  * "dsadsa"
  * 321
  * []

    (elixir) lib/regex.ex:397: Regex.split/3
```

Or repeat the function name:

```iex
iex(1)> Regex.split "dsadsa", 321
** (FunctionClauseError) there is no definition of Regex.split/3 that expects the following arguments:

  Regex.split("dsadsa", 321, [])

    (elixir) lib/regex.ex:397: Regex.split/3
```

Repeating the function name works well for small examples but it is problematic for large arguments.

This proposal is a small subset of #4171.

### PR DESCRIPTION
**Pull request by josevalim**
Closes #4171. Closes #5737.

Since `Exception.blame/3` is expensive, tools will have to explicitly opt-in to its usage when showing exceptions. In Elixir we will use this new functionality in:

  * [x] ExUnit
  * [x] IEx
  * [x] The exception message itself when blamed
  * [x] The compiler
  * [x] The CLI

Note: this will be included in Elixir v1.5 and it requires OTP 20 (currently RC but will be out before v1.5)

### CODE REVIEW

**Comment by josevalim** [comment]
Here is Plug using this feature to show what matched in a no function clause error:

screen shot 2017-05-21 at 21 50 33

**Comment by josevalim** [comment]
Now in ExUnit:

screen shot 2017-05-21 at 22 52 59

**Comment by josevalim** [comment]
In IEx (a bit too yellow for my taste):

screen shot 2017-05-22 at 13 21 27

**Comment by josevalim** [comment]
Hrm, I am thinking we need to highlight only where it fails, since by definition everything else succeeds.

**Comment by josevalim** [comment]
Here is a new IEx version (without the green bits):

screen shot 2017-05-22 at 13 47 58

**Comment by josevalim** [comment]
And ExUnit without the green bits too:

screen shot 2017-05-22 at 13 50 48

**Comment by josevalim** [comment]
This is ready! :heart:

File: lib/elixir/lib/exception.ex [hunk 1]
```diff
@@ -157,6 +157,159 @@ defmodule Exception do
...
+
+    * `FunctionClauseError` - annotated with the arguments
+      used on the call and available clauses
+
+  """
+  @spec blame(:error, any, stacktrace) :: t
+  @spec blame(kind, payload, stacktrace) :: payload when payload: var
+  def blame(:error, error, stacktrace) do
+    case normalize(:error, error, stacktrace) do
+      %{__struct__: FunctionClauseError} = struct ->
```
Comment (michalmuskala) [inline lib/elixir/lib/exception.ex]: Why not match on `%FunctionClauseError{}`?

Comment (josevalim) [inline lib/elixir/lib/exception.ex]: Bootstrapping. The struct is not yet available when we compile Exception.

File: lib/elixir/lib/exception.ex [hunk 2]
```diff
@@ -157,6 +157,159 @@ defmodule Exception do
...
+  defp rewrite_guard_call(:"/="), do: :!=
+  defp rewrite_guard_call(:"=:="), do: :===
+  defp rewrite_guard_call(:"=/="), do: :!==
+
+  defp rewrite_guard_call(op) when op in [:band, :bor, :bnot, :bsl, :bsr, :bxor],
+    do: {:., [], [Bitwise, op]}
+  defp rewrite_guard_call(op) when op in [:xor, :element, :size],
+    do: {:., [], [:erlang, op]}
+  defp rewrite_guard_call(op),
+    do: op
```
Comment (michalmuskala) [inline lib/elixir/lib/exception.ex]: Looking at this section, I can't avoid a feeling that something is off. This shouldn't be needed.

Why do we store this "expanded" version of the AST in debug info? Shouldn't we store the exact code before any such optimisations are performed?

Comment (josevalim) [inline lib/elixir/lib/exception.ex]: Excellent question which I asked myself during the process. If we store the original code, then we need to store the Macro.Env for every clause so we know how to expand it and that would make the :debug_info chunk very expensive in terms of size.

File: lib/elixir/lib/exception.ex [hunk 3]
```diff
@@ -157,6 +157,159 @@ defmodule Exception do
...
+  end
+
+  defp blame_args([call_arg | call_args], [ex_arg | ex_args], [erl_arg | erl_args], binding, ann_args) do
+    {match?, binding} = blame_arg(erl_arg, call_arg, binding)
+    blame_args(call_args, ex_args, erl_args, binding,
+               [blame_wrap(match?, rewrite_arg(ex_arg)) | ann_args])
+  end
+  defp blame_args([], [], [], binding, ann_args) do
+    {Enum.reverse(ann_args), binding}
+  end
```
Comment (whatyouhide) [inline lib/elixir/lib/exception.ex]: Just throwing it out there and didn't test at all (GitHub-comment-driven-development™), but would replacing the `{args, binding} = blame_args(call_args, ex_args, erl_args, [], [])` line above with the recursive function with a `zip` + `flat_map`? 

```elixir
[call_args, ex_args, erl_args]
|> Enum.zip()
|> Enum.map_reduce([], fn {call_arg, ex_arg, erl_arg}, binding -> some_function_name(call_arg, ex_arg, erl_arg, binding) end)

defp some_function_name(call_arg, ex_arg, erl_args, binding) do
  {match?, binding} = blame_arg(erl_arg, ex_arg, erl_args)
  {blame_wrap(match?, rewrite_arg(ex_arg)), binding}
end
```

?

File: lib/elixir/lib/exception.ex [hunk 4]
```diff
@@ -157,6 +157,160 @@ defmodule Exception do
...
+  This operation is potentially expensive, as it reads data
+  from the filesystem, parse beam files, evaluates code and
+  so on. Currently the following exceptions may be annotated:
+
+    * `FunctionClauseError` - annotated with the arguments
+      used on the call and available clauses
+
+  """
+  @spec blame(:error, any, stacktrace) :: t
+  @spec blame(kind, payload, stacktrace) :: payload when payload: var
```
Comment (fishcakez) [inline lib/elixir/lib/exception.ex]: The first argument has overlapping domain. Also the return is `{t, stacktrace} | {payload, stacktrace}`.

File: lib/elixir/lib/exception.ex [hunk 5]
```diff
@@ -157,6 +157,160 @@ defmodule Exception do
...
+  @doc """
+  Blames the invocation of the given module, function and arguments.
+
+  This function will retrieve the available clauses from bytecode
+  and evaluate them against the given arguments. The clauses are
+  returned as a list of `{args, guards}` pairs where each argument
+  and each top-level condition in a guard separated by `and`/`or`
+  is wrapped in a tuple with blame metadata.
+
+  This function returns either `{:ok, kind, clauses}` or `:error`.
```
Comment (fishcakez) [inline lib/elixir/lib/exception.ex]: Given `@type kind` in this module using `kind` for function definition kind is slightly confusing.

File: lib/iex/lib/iex/evaluator.ex [hunk 6]
```diff
@@ -248,11 +248,45 @@ defmodule IEx.Evaluator do
...
+    {blamed, stacktrace} = Exception.blame(kind, reason, stacktrace)
+    case blamed do
+      %FunctionClauseError{} ->
+        {_, inspect_opts} = pop_in IEx.inspect_opts[:syntax_colors][:reset]
+        Exception.format_banner(kind, reason, stacktrace) |> io_error
+        FunctionClauseError.blame(blamed, &inspect(&1, inspect_opts), &blame_match/2) |> io_blame
+      _ ->
+        Exception.format_banner(kind, blamed, stacktrace) |> io_error
+    end
+    Exception.format_stacktrace(prune_stacktrace(stacktrace)) |> io_stack
```
Comment (fishcakez) [inline lib/iex/lib/iex/evaluator.ex]: We should have single `IO` calls here too, though I see we previously didn't.

**Comment by josevalim** [comment]
All feedback has been incorporated. Thank you!

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**